File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/searching/hash_search.py
def hash_search(values: list[int], target: int) -> int:
    index_by_value: dict[int, int] = {}
    for index, value in enumerate(values):
        index_by_value.setdefault(value, index)
    return index_by_value.get(target, -1)


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 7, 99, 1]
    target = 19
    print("Hash-based Search")
    print(f"input : {sample}")
    print(f"target: {target}")
    print(f"index : {hash_search(sample, target)}")

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/searching/HashSearch.java
package com.sortingsearching.searching;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public final class HashSearch {
    private HashSearch() {
    }

    public static int search(int[] values, int target) {
        Map<Integer, Integer> indexByValue = new HashMap<>();
        for (int i = 0; i < values.length; i++) {
            indexByValue.putIfAbsent(values[i], i);
        }
        return indexByValue.getOrDefault(target, -1);
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 7, 99, 1};
        int target = 19;
        System.out.println("Hash-based Search");
        System.out.println("input : " + Arrays.toString(sample));
        System.out.println("target: " + target);
        System.out.println("index : " + search(sample, target));
    }
}
